<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Chris Simoes" />


<title>Calculating Relationship Matrices</title>

<script src="site_libs/header-attrs-2.6/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">lab5</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Calculating Relationship Matrices</h1>
<h4 class="author">Chris Simoes</h4>
<h4 class="date">3/12/2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-03-15
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>lab5/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20210314code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20210314)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20210314code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20210314)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongd4e4c77"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> d4e4c77 </a>
</p>
</div>
<div id="strongRepositoryversionstrongd4e4c77" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version d4e4c77. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .RData
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  analysis/Lab5_relationshipMatrices.Rmd
    Untracked:  code/addToREADME.R
    Untracked:  code/calcRelationshipMatrices.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="learning-objectives" class="section level2">
<h2>Learning Objectives</h2>
<ol style="list-style-type: decimal">
<li>Learn how to calculate pedigree and marker-based relationship matrices<br />
</li>
<li>Compare these relationships with the causal relationship matrix, considering the impact of the number of QTL<br />
</li>
<li>See if we can replicate a recent result from Naomi Wray and Peter Visscher`s group on differences in modeling phenotypic correlation for unrelated individuals versus close relatives</li>
</ol>
</div>
<div id="reproducibility" class="section level2">
<h2>Reproducibility</h2>
<p>Here is the <a href="./README.md">README file</a> for this report</p>
<div id="script-setup" class="section level3">
<h3>Script setup</h3>
<p>Install packages, start the README.md, set the random seed</p>
</div>
</div>
<div id="create-founder-haplotypes-and-set-trait" class="section level2">
<h2>Create founder haplotypes and set trait</h2>
<div id="genome-population-parameters" class="section level3">
<h3>Genome / Population parameters</h3>
<p>The historical effective population size (<code>historicalNe</code> here) is the size of the breeding population prior to recorded pedigrees.<br />
Test this script with <code>historicalNe</code> of either 100 or 1000.</p>
<pre class="r"><code>historicalNe &lt;- 800
nFounders &lt;- historicalNe
nChr &lt;- 7
segSites &lt;- 2000
nQTL &lt;- 200
nSNP &lt;- 1400
addToREADME(c(paste(&quot;Historical effective population size &quot;, historicalNe), paste(&quot;Number of founders &quot;, 
    nFounders), paste(&quot;Number of chromosomes &quot;, nChr), paste(&quot;Number of segregating sites per chromosome&quot;, 
    segSites), paste(&quot;Number of QTL per chromosome&quot;, nQTL), paste(&quot;Number of SNPs per chromosome&quot;, 
    nSNP)))</code></pre>
<blockquote>
<p>Think about the impact of the number of QTL on the QTL relationship matrix and how it compares to the genome-wide relationship matrix</p>
</blockquote>
</div>
<div id="set-trait" class="section level3">
<h3>Set trait</h3>
<p>Set the trait type and make founders. <code>runMacs2</code> gives more control over the population history of the simulated haplotypes. See documentation. Here, I want a simple constant effective population size, but not necessarily of 100, the default for runMacs.</p>
<pre class="r"><code>founderHaps &lt;- AlphaSimR::runMacs2(nInd = nFounders, nChr = nChr, segSites = segSites, 
    Ne = historicalNe, histNe = NULL, histGen = NULL)
SP &lt;- AlphaSimR::SimParam$new(founderHaps)
SP$restrSegSites(minQtlPerChr = nQTL, minSnpPerChr = nSNP, overlap = FALSE)  # This prevents SNP from also being QTL
SP$addTraitA(nQtlPerChr = nQTL)
SP$addSnpChip(nSnpPerChr = nSNP)

# Create a new population of founders
founders &lt;- AlphaSimR::newPop(founderHaps, simParam = SP)</code></pre>
</div>
<div id="run-generations-to-create-pedigree-setup" class="section level3">
<h3>Run generations to create pedigree: setup</h3>
<p>Specify generation sizes here</p>
<pre class="r"><code># Number of selection cycles
nCyc &lt;- 5
# Narrow-sense heritability
h2 &lt;- 0.4
# Allow for a different number of progeny over cycles
# Make last gen bigger for sample size
nProgeny &lt;- rep(500, nCyc)
nProgeny[nCyc] &lt;- 1000
# Whether to randomly mate or to select (that is the question!)
randomOrSelect &lt;- &quot;random&quot;  # Or put &#39;select&#39; if you want...
# Set the number to select. If nToSelect == nProgeny, it&#39;s random mating
nToSelect &lt;- nProgeny
if (randomOrSelect == &quot;select&quot;) nToSelect &lt;- nToSelect/10

addToREADME(c(&quot;Chunk Run generations setup&quot;, paste(&quot;Number of cycles&quot;, nCyc), paste(&quot;Heritability&quot;, 
    h2), paste(&quot;Random mating or selection&quot;, randomOrSelect), paste(&quot;Number of progeny&quot;, 
    paste(nProgeny, collapse = &quot;, &quot;)), paste(&quot;Number to select&quot;, paste(nToSelect, 
    collapse = &quot;, &quot;))))

# Matrix to save pedigrees
pedRec &lt;- cbind(id = founders@id, seedPar = founders@mother, pollenPar = founders@father)</code></pre>
</div>
<div id="run-generations-to-create-pedigree-loop" class="section level3">
<h3>Run generations to create pedigree: loop</h3>
<p><code>AlphaSimR</code> creates the founders with a coalescent and therefore no pedigree. So we can`t look at a pedigree-based relationship matrix on the founders. Here we advance the founders a few generations just to create that pedigree.</p>
<pre class="r"><code>selected &lt;- founders
for (cyc in 1:nCyc) {
    # Create new population by random mating
    # nCrosses: how many crosses to make
    # nProgeny: number of progeny per cross
    progenyPop &lt;- AlphaSimR::randCross(selected, nCrosses = nProgeny[cyc], nProgeny = 1)
    # Get phenotypes
    progenyPop &lt;- AlphaSimR::setPheno(progenyPop, h2 = h2)
    # Store pedigree in pedRec for analysis
    pedRec &lt;- rbind(pedRec, cbind(id = progenyPop@id, seedPar = progenyPop@mother, 
        pollenPar = progenyPop@father))
    # Select
    # If randomOrSelect == &#39;random&#39;, nToSelect will be size of pop: no selection
    if (cyc &lt; nCyc) 
        selected &lt;- AlphaSimR::selectInd(progenyPop, nToSelect[cyc])
}</code></pre>
</div>
<div id="calculate-additive-relationship-matrices" class="section level3">
<h3>Calculate additive relationship matrices</h3>
<p>The true or causal relationship matrix is that of QTL. In reality, we don<code>t have access to that matrix and we approximated it with the pedigree- or SNP-based relationship matrix.   The pedigree-based relationship matrix is the expectation of the QTL relationship matrix. The QTL relationship matrix is one realization of that expectation.   The SNP relationship matrix is another realization. Assuming SNPs are in LD with the QTL, it should be more strongly correlated to the QTL matrix than the pedigree-based matrix.   **NOTE** The code to make these matrices is in "code/calcRelationshipMatrices.R" That</code>s the meat of this lab, so we will have a look.</p>
<pre class="r"><code>source(here::here(&quot;code/calcRelationshipMatrices.R&quot;))
qtlMat &lt;- AlphaSimR::pullQtlGeno(progenyPop)
qtlRelMat &lt;- calcGenomicRelationshipMatrix(qtlMat)

snpMat &lt;- AlphaSimR::pullSnpGeno(progenyPop)
snpRelMat &lt;- calcGenomicRelationshipMatrix(snpMat)

numPedMat &lt;- convertNamesToRows(pedRec)  # AlphaSimR uses character names for id
aMat &lt;- 2 * pedigreeToCCmatrix(numPedMat)
# Just keep the aMat for the final generation
totProg &lt;- nrow(numPedMat)
toKeep &lt;- totProg + (1 - nProgeny[cyc]):0
aMat &lt;- aMat[toKeep, toKeep]</code></pre>
<pre class="r"><code># Regress of diagonal elements of one matrix on another
# Function to return offdiagonal elements of a relationship matrix
offDiag &lt;- function(relMat) {
    return(c(relMat[upper.tri(relMat)]))
}</code></pre>
<pre class="r"><code># This bit only works if Ne == 100 or Ne == 1000
# I have set bin sizes so that these work nicely for those two population sizes
# To try a different population size, figure out how to bin n*(n-1)/2
if (nProgeny[cyc] == 100) binSizes &lt;- c(75, 200, 200, rep(500, 8), 200, 200, 75) else binSizes &lt;- c(250, 
    1000, 1000, 2500, rep(10000, 49), 2500, 1000, 1000, 250)

# Cheating a bit here and using genotypic values rather than phenotypes
stdPheno &lt;- scale(AlphaSimR::gv(progenyPop), center = T, scale = T)

# Given the number of pairs that you want in each bin, calculate where to place the bin boundaries
calcBinBoundaries &lt;- function(dataVec, binSizes) {
    csbs &lt;- cumsum(binSizes)
    binPerc &lt;- csbs/max(csbs)
    return(quantile(dataVec, probs = binPerc))
}

# Calculate the mean relatedness and phenotypic correlation within bins 
calcBinParms &lt;- function(bin, relMat) {
    binBound &lt;- calcBinBoundaries(offDiag(relMat), binSizes)
    if (bin == 1) 
        pairs &lt;- which(relMat &lt; binBound[bin] &amp; upper.tri(relMat), arr.ind = T) else pairs &lt;- which(relMat &gt;= binBound[bin - 1] &amp; relMat &lt; binBound[bin] &amp; upper.tri(relMat), 
        arr.ind = T)
    pairCor &lt;- stdPheno[pairs[, 1]] * stdPheno[pairs[, 2]]
    return(c(mean(relMat[pairs]), mean(pairCor)))
}

whichMat &lt;- &quot;SNP&quot;
relMat &lt;- switch(whichMat, Pedigree = aMat, QTL = qtlRelMat, SNP = snpRelMat, )
binParms &lt;- t(sapply(1:length(binSizes), calcBinParms, relMat = relMat))
xlim &lt;- range(binParms[, 1])
ylim &lt;- range(binParms[, 2])
plot(binParms, xlab = paste(whichMat, &quot;relatedness&quot;), ylab = &quot;Phenotypic correlation&quot;, 
    pch = 16)
lines(range(binParms), range(binParms), col = &quot;red&quot;)

whichMat &lt;- &quot;QTL&quot;
relMat &lt;- switch(whichMat, Pedigree = aMat, QTL = qtlRelMat, SNP = snpRelMat, )
binParms &lt;- t(sapply(1:length(binSizes), calcBinParms, relMat = relMat))
xlim &lt;- range(binParms[, 1])
ylim &lt;- range(binParms[, 2])
plot(binParms, xlab = paste(whichMat, &quot;relatedness&quot;), ylab = &quot;Phenotypic correlation&quot;, 
    pch = 16)
lines(range(binParms), range(binParms), col = &quot;red&quot;)</code></pre>
<p><img src="figure/Lab5_relationshipMatrices.Rmd/Imitate%20Visscher-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/Lab5_relationshipMatrices.Rmd/Imitate%20Visscher-2.png" width="672" style="display: block; margin: auto;" /></p>
<blockquote>
<p>What conditions of QTL number and effective population size lead to a relationship between phenotypic correlation and SNP relatedness similar to that observed in Kemper et al. 2021?</p>
</blockquote>
<p>Note that if you set <code>historicalNe</code> much bigger than 1000, it’s going to take some time…</p>
<p>Note that the <code>echo = FALSE</code> parameter was added to the code chunk to prevent printing of the R code that generated the plot.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6       rstudioapi_0.13  knitr_1.30       magrittr_2.0.1  
 [5] workflowr_1.6.2  hms_0.5.3        here_1.0.1       R6_2.5.0        
 [9] rlang_0.4.10     AlphaSimR_0.13.0 stringr_1.4.0    tools_4.0.3     
[13] xfun_0.20        git2r_0.28.0     htmltools_0.5.1  ellipsis_0.3.1  
[17] yaml_2.2.1       digest_0.6.27    rprojroot_2.0.2  tibble_3.0.4    
[21] lifecycle_0.2.0  crayon_1.3.4     formatR_1.7      readr_1.4.0     
[25] later_1.1.0.1    vctrs_0.3.6      promises_1.1.1   fs_1.5.0        
[29] glue_1.4.2       evaluate_0.14    rmarkdown_2.6    stringi_1.5.3   
[33] compiler_4.0.3   pillar_1.4.7     httpuv_1.5.4     pkgconfig_2.0.3 </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
